dsumaroka/jv-stream-practice

dsumaroka opened the pull request.

olekskov commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -23,7 +39,21 @@ public int findMinEvenNumber(List<String> numbers) {
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        Stream<Integer> streamOdd = IntStream.range(0, numbers.size())
+                .filter(i -> i % 2 != 0)
+                .mapToObj(numbers::get)
+                .map(n -> n - 1);
+
+        Stream<Integer> streamEven = IntStream.range(0, numbers.size())
+                .filter(i -> i % 2 == 0)
+                .mapToObj(numbers::get);
+
+        return Stream.concat(streamEven, streamOdd)
+                .filter(n -> n % 2 != 0)
+                .mapToDouble(n -> n)
+                .average()
+                .orElseThrow();
```
Please try to make in one stream. You're correct that we need to use range.
But next we can use .mapToDouble method with ternary operator inside to get element from list or element - 1 to combine a lot of actions

olekskov commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +20,17 @@ public class StreamPractice {
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        try {
+            return numbers.stream()
+                    .map(s -> s.split(","))
+                    .flatMap(Arrays::stream)
+                    .mapToInt(Integer::valueOf)
+                    .filter(n -> n % 2 == 0)
+                    .reduce(Integer::min)
+                    .getAsInt();
```
We shouldn't use 1-2 letter abbreviations for variables even in stream. Lines are already short, normal full naming will only make it more readable

olekskov replied:
```suggestion
                    .min()
                    .orElseThrow(...some proper exception here..);
```

olekskov commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +20,17 @@ public class StreamPractice {
-        return 0;
+        try {
+            return numbers.stream()
+                    .map(s -> s.split(","))
+                    .flatMap(Arrays::stream)
```
map+flatmap usually can be replaced with just flatmap
.flatMap(s -> Arrays.stream(s.split(",")))

olekskov commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +20,17 @@ public class StreamPractice {
+                    .filter(n -> n % 2 == 0)
+                    .reduce(Integer::min)
+                    .getAsInt();
+        } catch (NoSuchElementException e) {
+            throw new NoSuchElementException("Can't get min value from list");
```
and then try/catch is not required at all

olekskov commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -59,7 +96,12 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex().equals(Person.Sex.WOMAN) && p.getAge() >= femaleAge)
+                .map(Person::getCats)
+                .flatMap(Collection::stream)
```
map+flatmap -> one proper flatmap

dsumaroka commented:
Ок

Пн, 29 янв. 2024 г. в 21:26, Oleksandr Kovalenko ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470086310>
> :
>
> > +        Stream<Integer> streamOdd = IntStream.range(0, numbers.size())
> +                .filter(i -> i % 2 != 0)
> +                .mapToObj(numbers::get)
> +                .map(n -> n - 1);
> +
> +        Stream<Integer> streamEven = IntStream.range(0, numbers.size())
> +                .filter(i -> i % 2 == 0)
> +                .mapToObj(numbers::get);
> +
> +        return Stream.concat(streamEven, streamOdd)
> +                .filter(n -> n % 2 != 0)
> +                .mapToDouble(n -> n)
> +                .average()
> +                .orElseThrow();
>
> Please try to make in one stream. You're correct that we need to use range.
> But next we can use .mapToDouble method with ternary operator inside to
> get element from list or element - 1 to combine a lot of actions
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470087867>
> :
>
> > +            return numbers.stream()
> +                    .map(s -> s.split(","))
> +                    .flatMap(Arrays::stream)
> +                    .mapToInt(Integer::valueOf)
> +                    .filter(n -> n % 2 == 0)
> +                    .reduce(Integer::min)
> +                    .getAsInt();
>
> We shouldn't use 1-2 letter abbreviations for variables even in stream.
> Lines are already short, normal full naming will only make it more readable
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470093444>
> :
>
> > +                    .map(s -> s.split(","))
> +                    .flatMap(Arrays::stream)
>
> map+flatmap usually can be replaced with just flatmap
> .flatMap(s -> Arrays.stream(s.split(",")))
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470094071>
> :
>
> > +                    .reduce(Integer::min)
> +                    .getAsInt();
>
> ⬇️ Suggested change
>
> -                    .reduce(Integer::min)
> -                    .getAsInt();
> +                    .min()
> +                    .orElseThrow(...some proper exception here..);
>
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470094924>
> :
>
> > +        } catch (NoSuchElementException e) {
> +            throw new NoSuchElementException("Can't get min value from list");
>
> and then try/catch is not required at all
> ------------------------------
>
> In src/main/java/practice/StreamPractice.java
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#discussion_r1470095448>
> :
>
> > +                .map(Person::getCats)
> +                .flatMap(Collection::stream)
>
> map+flatmap -> one proper flatmap
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-stream-github-practice/pull/997#pullrequestreview-1849584821>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7UEH4GXAZ2F2QDA6AX57OTYQ7ZX7AVCNFSM6AAAAABCPZIYEWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNBZGU4DIOBSGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>
